### PR TITLE
CMake improvements

### PR DESCRIPTION
**Pull request by moien007**
This is PR https://github.com/lsalzman/enet/pull/137 (which I messed up, my bad) plus new changes.

Changes are:
- Upgrading from CMake 2.6 to 3.4.3
- Linking against `ws2_32` and `winmm` if compiler is MSVC
- Conditionally defining `ENET_DLL` macro
- Exporting all symbols for Win32 so that a `.def` file or `__declspec(dllexport)` is not required

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -18,8 +18,14 @@ check_struct_has_member("struct msghdr" "msg_flags" "sys/types.h;sys/socket.h" H
...
 unset(CMAKE_EXTRA_INCLUDE_FILES)
+
+
+if (BUILD_SHARED_LIBS)
+    add_definitions(-DENET_SHARED_LIB=1)
+endif()
+
 if(MSVC)
-	add_definitions(-W3)
+add_definitions(-W3)
```
Comment (JayFoxRox) [inline CMakeLists.txt:28]: Unintentional whitespace change?

File: CMakeLists.txt [hunk 2]
```diff
@@ -84,11 +90,15 @@ set(SOURCE_FILES
...
-add_library(enet STATIC
+add_library(enet
     ${INCLUDE_FILES}
     ${SOURCE_FILES}
 )
 
-if (MINGW)
+if (WIN32)
+    if (BUILD_SHARED_LIBS)
+        set_property(TARGET enet PROPERTY WINDOWS_EXPORT_ALL_SYMBOLS ON)
```
Comment (JayFoxRox) [inline CMakeLists.txt:100]: I've never used `WINDOWS_EXPORT_ALL_SYMBOLS`, but it sounds bad.
I'd be concerned with polluting the namespace; I personally prefer explicit exports of the symbols that should be exported.

Comment (moien007) [reply CMakeLists.txt:100]: I am aware of CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS but `enet` does not mark its API's with dllexport attribute.
We can add .def which is a pain to maintain or add dllexport attribute which I think is a good alternative option.

Comment (moien007) [reply CMakeLists.txt:100]: Notice that this property only affects Win32 DLL files. A enet.dll file on windows with all symbols exported and compiled with `/MT` is about 700kb which I don't find big.

File: CMakeLists.txt (hunk 1 above)
Comment (moien007) [reply CMakeLists.txt:28]: yes.

**Pull request closed without merging**